Implement Streaming Response Support for LLM Providers

The current LLM provider implementation only supports single-response completions, which can lead to long wait times for users, especially with longer prompts or analyses. Adding streaming support would significantly improve the user experience by providing immediate feedback and partial results as they become available.
Key implementation points:
1. Update ILLMProvider interface to add streaming method:
```csharp
Task<IAsyncEnumerable<string>> GetCompletionStreamAsync(string prompt);
```
2. Modify BaseLLMProvider to support streaming:
- Add abstract streaming method
- Implement caching for streamed responses
- Ensure rate limiting works with streaming
- Add configuration option to enable/disable streaming
3. Update Anthropic and OpenAI clients to implement streaming:
- Use proper API endpoints for streaming
- Handle chunked responses correctly
- Maintain connection stability
- Add proper error handling for stream interruptions
4. Update CLI interface:
- Add progress indicators for streaming responses
- Allow users to configure streaming preference
- Implement proper terminal handling for streamed output
- Add support for cancelling ongoing streams
This improvement will enhance user experience, reduce perceived latency, and bring the tool in line with modern AI interaction patterns. It's particularly important given the tool's interactive nature and the potential for long-form code analysis responses.

⚠️ Build Failure

The changes I made resulted in build failures:

```
/app/repository/tests/XPlain.Tests/XPlain.Tests.csproj(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.

Build FAILED.

/app/repository/tests/XPlain.Tests/XPlain.Tests.csproj(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.21

```

I'll review and fix these build errors before proceeding.

--- a/src/XPlain/Configuration/CommandLineOptions.cs
+++ b/src/XPlain/Configuration/CommandLineOptions.cs
@@ -81,6 +81,10 @@ public class CommandLineOptions
     [Description("Display cache statistics")]
     public bool ShowCacheStats { get; set; }
 
+    [OptionGroup(OptionGroup.Output)]
+    [Description("Enable streaming responses")]
+    public bool EnableStreaming { get; set; } = false;
+
     public bool InteractiveMode => string.IsNullOrEmpty(DirectQuestion) && !ShowCacheStats;
 
     public void Validate()

--- a/src/XPlain/Configuration/StreamingSettings.cs
+++ b/src/XPlain/Configuration/StreamingSettings.cs
@@ -1,0 +1,24 @@
+namespace XPlain.Configuration;
+
+public class StreamingSettings
+{
+    /// <summary>
+    /// Whether streaming is enabled by default
+    /// </summary>
+    public bool EnableStreamingByDefault { get; set; } = false;
+
+    /// <summary>
+    /// Maximum timeout for streaming connections in seconds
+    /// </summary>
+    public int StreamingTimeoutSeconds { get; set; } = 30;
+
+    /// <summary>
+    /// Number of retry attempts for failed streaming connections
+    /// </summary>
+    public int MaxStreamingRetries { get; set; } = 3;
+
+    /// <summary>
+    /// Initial delay between retries in milliseconds
+    /// </summary>
+    public int InitialRetryDelayMs { get; set; } = 1000;
+}

--- a/src/XPlain/Program.cs
+++ b/src/XPlain/Program.cs
@@ -635,6 +635,10 @@ file class Program
                     catch (Exception ex)
                     {
                         Console.Error.WriteLine($"Error in command processing: {ex.Message}");
+                        if (options.VerbosityLevel >= 2)
+                        {
+                            Console.WriteLine($"Full error: {ex}");
+                        }
                     }
                     finally
                     {
@@ -675,6 +679,10 @@ file class Program
             Console.WriteLine("Navigation:");
             Console.WriteLine("  exit     - Exit the application");
             Console.WriteLine("  quit     - Exit the application");
+            Console.WriteLine();
+            Console.WriteLine("Features:");
+            Console.WriteLine("  - Streaming responses enabled: " + (CommandLineOptions.EnableStreaming ? "Yes" : "No"));
+            Console.WriteLine("  - Use --enable-streaming flag to see responses in real-time");
             Console.WriteLine();
             Console.WriteLine("Type your questions about the code to analyze them.");
         }
@@ -789,8 +797,8 @@ file class Program
                               # Find specific information
                               xplain ./my-project -q "List all interfaces and their implementations"
 
-                              Output Formats
-                              -------------
+                              Output and Interaction
+                              --------------------
                               # Markdown for documentation
                               xplain ./my-project -f markdown -q "Generate API documentation"
 
@@ -799,6 +807,9 @@ file class Program
 
                               # Plain text (default)
                               xplain ./my-project -q "Explain the error handling"
+
+                              # Streaming responses (real-time output)
+                              xplain ./my-project --enable-streaming -q "Analyze the architecture"
 
                               Verbosity Levels
                               ---------------

--- a/src/XPlain/Program.cs
+++ b/src/XPlain/Program.cs
@@ -653,14 +653,15 @@ file class Program
                                     {
                                         await using var responseStream = provider.GetCompletionStreamAsync(prompt);
                                         var isFirstChunk = true;
-                                        var progressTimer = new Timer(_ => UpdateProgressIndicator(), null, 0, 100);
+                                        
+                                        using var progress = new ProgressIndicator();
+                                        progress.Start();
                                         
                                         await foreach (var chunk in responseStream.WithCancellation(cts.Token))
                                         {
                                             if (isFirstChunk)
                                             {
-                                                progressTimer.Dispose();
-                                                Console.Write("\b \b"); // Clear progress indicator
+                                                progress.Stop();
                                                 isFirstChunk = false;
                                             }
                                             Console.Write(chunk);

--- a/src/XPlain/Program.cs
+++ b/src/XPlain/Program.cs
@@ -412,8 +412,22 @@ file class Program
                 else
                 {
                     string codeContext = BuildCodeContext(options.CodebasePath);
-                    string response = await provider.GetCompletionAsync($"I have the following code:\n\n{codeContext}\n\nMy question is: {options.DirectQuestion}");
-                    OutputResponse(response, options.OutputFormat);
+                    string prompt = $"I have the following code:\n\n{codeContext}\n\nMy question is: {options.DirectQuestion}";
+
+                    if (options.EnableStreaming)
+                    {
+                        await using var responseStream = provider.GetCompletionStreamAsync(prompt);
+                        await foreach (var chunk in responseStream)
+                        {
+                            Console.Write(chunk);
+                        }
+                        Console.WriteLine(); // Add newline after streaming response
+                    }
+                    else
+                    {
+                        string response = await provider.GetCompletionAsync(prompt);
+                        OutputResponse(response, options.OutputFormat);
+                    }
                 }
 
                 return 0;
@@ -577,13 +591,32 @@ file class Program
                             try
                             {
                                 string codeContext = BuildCodeContext(options.CodebasePath);
-                                string response = await provider.GetCompletionAsync($"I have the following code:\n\n{codeContext}\n\nMy question is: {input}");
-                                if (options.VerbosityLevel >= 1)
-                                {
-                                    Console.WriteLine("\nResponse:");
-                                }
+                                string prompt = $"I have the following code:\n\n{codeContext}\n\nMy question is: {input}";
 
-                                OutputResponse(response, options.OutputFormat);
+                                if (options.EnableStreaming)
+                                {
+                                    if (options.VerbosityLevel >= 1)
+                                    {
+                                        Console.WriteLine("\nStreaming response:");
+                                    }
+
+                                    await using var responseStream = provider.GetCompletionStreamAsync(prompt);
+                                    await foreach (var chunk in responseStream)
+                                    {
+                                        Console.Write(chunk);
+                                    }
+                                    Console.WriteLine(); // Add newline after streaming response
+                                }
+                                else
+                                {
+                                    string response = await provider.GetCompletionAsync(prompt);
+                                    if (options.VerbosityLevel >= 1)
+                                    {
+                                        Console.WriteLine("\nResponse:");
+                                    }
+
+                                    OutputResponse(response, options.OutputFormat);
+                                }
                             }
                             catch (Exception ex)
                             {

--- a/src/XPlain/Program.cs
+++ b/src/XPlain/Program.cs
@@ -18,6 +18,19 @@ file class Program
     {
         private const string Version = "1.0.0";
         static bool _keepRunning = true;
+        private static readonly string[] ProgressIndicators = new[] { "⣾", "⣽", "⣻", "⢿", "⡿", "⣟", "⣯", "⣷" };
+        private static int _progressIndex = 0;
+        
+        private static void UpdateProgressIndicator()
+        {
+            if (Console.IsOutputRedirected) return;
+            
+            var currentPos = Console.CursorLeft;
+            Console.Write("\b \b"); // Clear previous indicator
+            Console.Write(ProgressIndicators[_progressIndex]);
+            Console.CursorLeft = currentPos;
+            _progressIndex = (_progressIndex + 1) % ProgressIndicators.Length;
+        }
 
         internal static async Task ProcessCommandInternal(CommandLineOptions options)
         {
@@ -416,12 +429,41 @@ file class Program
 
                     if (options.EnableStreaming)
                     {
-                        await using var responseStream = provider.GetCompletionStreamAsync(prompt);
-                        await foreach (var chunk in responseStream)
+                        using var cts = new CancellationTokenSource();
+                        Console.CancelKeyPress += (s, e) => {
+                            e.Cancel = true; // Prevent process termination
+                            cts.Cancel();
+                            Console.WriteLine("\nCancelling stream...");
+                        };
+
+                        try 
                         {
-                            Console.Write(chunk);
+                            if (options.VerbosityLevel >= 1)
+                            {
+                                Console.WriteLine("Streaming response (Ctrl+C to cancel):");
+                            }
+
+                            await using var responseStream = provider.GetCompletionStreamAsync(prompt);
+                            var isFirstChunk = true;
+                            var progressTimer = new Timer(_ => UpdateProgressIndicator(), null, 0, 100);
+                            
+                            await foreach (var chunk in responseStream.WithCancellation(cts.Token))
+                            {
+                                if (isFirstChunk)
+                                {
+                                    progressTimer.Dispose();
+                                    Console.Write("\b \b"); // Clear progress indicator
+                                    isFirstChunk = false;
+                                }
+                                Console.Write(chunk);
+                            }
+                            Console.WriteLine(); // Add newline after streaming response
                         }
-                        Console.WriteLine(); // Add newline after streaming response
+                        catch (OperationCanceledException)
+                        {
+                            Console.WriteLine("\nResponse stream cancelled.");
+                            return 1;
+                        }
                     }
                     else
                     {
@@ -597,15 +639,38 @@ file class Program
                                 {
                                     if (options.VerbosityLevel >= 1)
                                     {
-                                        Console.WriteLine("\nStreaming response:");
+                                        Console.WriteLine("\nStreaming response (Ctrl+C to cancel):");
                                     }
 
-                                    await using var responseStream = provider.GetCompletionStreamAsync(prompt);
-                                    await foreach (var chunk in responseStream)
+                                    using var cts = new CancellationTokenSource();
+                                    Console.CancelKeyPress += (s, e) => {
+                                        e.Cancel = true; // Prevent process termination
+                                        cts.Cancel();
+                                        Console.WriteLine("\nCancelling stream...");
+                                    };
+
+                                    try 
                                     {
-                                        Console.Write(chunk);
+                                        await using var responseStream = provider.GetCompletionStreamAsync(prompt);
+                                        var isFirstChunk = true;
+                                        var progressTimer = new Timer(_ => UpdateProgressIndicator(), null, 0, 100);
+                                        
+                                        await foreach (var chunk in responseStream.WithCancellation(cts.Token))
+                                        {
+                                            if (isFirstChunk)
+                                            {
+                                                progressTimer.Dispose();
+                                                Console.Write("\b \b"); // Clear progress indicator
+                                                isFirstChunk = false;
+                                            }
+                                            Console.Write(chunk);
+                                        }
+                                        Console.WriteLine(); // Add newline after streaming response
                                     }
-                                    Console.WriteLine(); // Add newline after streaming response
+                                    catch (OperationCanceledException)
+                                    {
+                                        Console.WriteLine("\nResponse stream cancelled.");
+                                    }
                                 }
                                 else
                                 {
@@ -679,10 +744,12 @@ file class Program
             Console.WriteLine("Navigation:");
             Console.WriteLine("  exit     - Exit the application");
             Console.WriteLine("  quit     - Exit the application");
+            Console.WriteLine("  ctrl+c   - Cancel current streaming response");
             Console.WriteLine();
             Console.WriteLine("Features:");
             Console.WriteLine("  - Streaming responses enabled: " + (CommandLineOptions.EnableStreaming ? "Yes" : "No"));
             Console.WriteLine("  - Use --enable-streaming flag to see responses in real-time");
+            Console.WriteLine("  - Press Ctrl+C to cancel streaming responses");
             Console.WriteLine();
             Console.WriteLine("Type your questions about the code to analyze them.");
         }

--- a/src/XPlain/Services/AnthropicClient.cs
+++ b/src/XPlain/Services/AnthropicClient.cs
@@ -41,7 +41,11 @@ public class AnthropicClient : BaseLLMProvider, IAnthropicClient, IDisposable
         _settings = settings.Value;
         _rateLimitingService = rateLimitingService;
         
-        _httpClient = new HttpClient
+        _httpClient = new HttpClient(
+            new StreamingHttpHandler(
+                timeout: TimeSpan.FromSeconds(30),
+                maxRetries: 3,
+                initialRetryDelay: TimeSpan.FromSeconds(1)))
         {
             BaseAddress = new Uri(_settings.ApiEndpoint)
         };
@@ -166,7 +170,7 @@ public class AnthropicClient : BaseLLMProvider, IAnthropicClient, IDisposable
         try
         {
             using var response = await _httpClient.PostAsJsonAsync(
-                "/v1/messages",
+                "/v1/messages/stream",
                 request,
                 cancellationToken);
 

--- a/src/XPlain/Services/AnthropicClient.cs
+++ b/src/XPlain/Services/AnthropicClient.cs
@@ -1,10 +1,13 @@
 using System.Diagnostics;
 using System.Net.Http.Headers;
 using System.Net.Http.Json;
+using System.Text.Json;
 using System.Text.Json.Serialization;
 using Microsoft.Extensions.Options;
 using XPlain.Configuration;
 using System.Net;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
 
 namespace XPlain.Services;
 
@@ -73,6 +76,126 @@ public class AnthropicClient : BaseLLMProvider, IAnthropicClient, IDisposable
             return response?.Content.FirstOrDefault()?.Text.Trim() ?? 
                    "No response received from the API.";
         }, 0);
+    }
+
+    protected override async IAsyncEnumerable<string> GetCompletionStreamInternalAsync(
+        string prompt)
+    {
+        var requestBody = new AnthropicRequest
+        {
+            Model = _settings.DefaultModel,
+            Messages = [
+                new AnthropicMessage
+                {
+                    Role = "user",
+                    Content = [new AnthropicMessageContent { Type = "text", Text = prompt }]
+                }
+            ],
+            MaxTokens = _settings.MaxTokenLimit,
+            Temperature = 0.7,
+            Stream = true
+        };
+
+        await foreach (var chunk in ExecuteStreamWithRetryAsync(requestBody))
+        {
+            yield return chunk;
+        }
+    }
+
+    private async IAsyncEnumerable<string> ExecuteStreamWithRetryAsync(
+        AnthropicRequest request,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        if (!_circuitBreaker.CanProcess())
+        {
+            throw new CircuitBreakerOpenException(
+                "Circuit breaker is open. Too many recent failures.");
+        }
+
+        var attempts = 0;
+        var delay = _settings.InitialRetryDelayMs;
+        Exception lastException = null;
+
+        while (attempts < _settings.MaxRetryAttempts)
+        {
+            try
+            {
+                attempts++;
+                await foreach (var chunk in ProcessStreamRequestAsync(request, cancellationToken))
+                {
+                    yield return chunk;
+                }
+                _circuitBreaker.RecordSuccess();
+                yield break;
+            }
+            catch (HttpRequestException ex) when (
+                ex.StatusCode.HasValue &&
+                RetryableStatusCodes.Contains(ex.StatusCode.Value))
+            {
+                lastException = ex;
+                _circuitBreaker.RecordFailure();
+
+                if (attempts < _settings.MaxRetryAttempts)
+                {
+                    var jitter = Random.Shared.NextDouble() * _settings.JitterFactor * delay;
+                    await Task.Delay((int)(delay + jitter), cancellationToken);
+                    delay = Math.Min(
+                        (int)(delay * _settings.BackoffMultiplier),
+                        _settings.MaxRetryDelayMs);
+                }
+            }
+            catch (Exception ex)
+            {
+                lastException = ex;
+                _circuitBreaker.RecordFailure();
+                throw;
+            }
+        }
+
+        throw new AnthropicApiException(
+            $"Stream operation failed after {attempts} attempts. Last error: {lastException?.Message}",
+            lastException);
+    }
+
+    private async IAsyncEnumerable<string> ProcessStreamRequestAsync(
+        AnthropicRequest request,
+        [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        // Acquire rate limiting permit
+        await _rateLimitingService.AcquirePermitAsync(ProviderName, 0, cancellationToken);
+        try
+        {
+            using var response = await _httpClient.PostAsJsonAsync(
+                "/v1/messages",
+                request,
+                cancellationToken);
+
+            response.EnsureSuccessStatusCode();
+            
+            using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
+            using var reader = new StreamReader(stream);
+
+            while (!reader.EndOfStream)
+            {
+                var line = await reader.ReadLineAsync();
+                if (string.IsNullOrEmpty(line)) continue;
+                if (!line.StartsWith("data: ")) continue;
+                if (line == "data: [DONE]") break;
+
+                var jsonData = line.Substring(6);
+                var streamResponse = JsonSerializer.Deserialize<AnthropicStreamResponse>(jsonData);
+                
+                if (streamResponse?.Type == "content_block_delta" && 
+                    !string.IsNullOrEmpty(streamResponse.Delta?.Text))
+                {
+                    yield return streamResponse.Delta.Text;
+                }
+            }
+        }
+        finally
+        {
+            _rateLimitingService.ReleasePermit(ProviderName);
+        }
     }
 
     private async Task<AnthropicResponse> ProcessRequestAsync(
@@ -269,4 +392,25 @@ public class AnthropicResponse
 
     [JsonPropertyName("content")]
     public required List<AnthropicMessageContent> Content { get; set; }
+}
+
+public class AnthropicStreamResponse
+{
+    [JsonPropertyName("type")]
+    public required string Type { get; set; }
+
+    [JsonPropertyName("delta")]
+    public AnthropicDelta? Delta { get; set; }
+
+    [JsonPropertyName("index")]
+    public int Index { get; set; }
+}
+
+public class AnthropicDelta
+{
+    [JsonPropertyName("type")]
+    public string? Type { get; set; }
+
+    [JsonPropertyName("text")]
+    public string? Text { get; set; }
 }

--- a/src/XPlain/Services/AnthropicClient.cs
+++ b/src/XPlain/Services/AnthropicClient.cs
@@ -35,17 +35,14 @@ public class AnthropicClient : BaseLLMProvider, IAnthropicClient, IDisposable
     public AnthropicClient(
         IOptions<AnthropicSettings> settings,
         ICacheProvider cacheProvider,
-        IRateLimitingService rateLimitingService)
-        : base(cacheProvider, rateLimitingService)
+        IRateLimitingService rateLimitingService,
+        StreamingSettings streamingSettings)
+        : base(cacheProvider, rateLimitingService, streamingSettings)
     {
         _settings = settings.Value;
         _rateLimitingService = rateLimitingService;
         
-        _httpClient = new HttpClient(
-            new StreamingHttpHandler(
-                timeout: TimeSpan.FromSeconds(30),
-                maxRetries: 3,
-                initialRetryDelay: TimeSpan.FromSeconds(1)))
+        _httpClient = new HttpClient(new StreamingHttpHandler(streamingSettings))
         {
             BaseAddress = new Uri(_settings.ApiEndpoint)
         };

--- a/src/XPlain/Services/BaseLLMProvider.cs
+++ b/src/XPlain/Services/BaseLLMProvider.cs
@@ -1,5 +1,7 @@
+using System.Collections.Generic;
 using System.Security.Cryptography;
 using System.Text;
+using System.Runtime.CompilerServices;
 
 namespace XPlain.Services;
 
@@ -20,6 +22,8 @@ public abstract class BaseLLMProvider : ILLMProvider
     public abstract string ModelName { get; }
 
     protected abstract Task<string> GetCompletionInternalAsync(string prompt);
+    
+    protected abstract IAsyncEnumerable<string> GetCompletionStreamInternalAsync(string prompt);
 
     public async Task<string> GetCompletionAsync(string prompt)
     {
@@ -48,6 +52,52 @@ public abstract class BaseLLMProvider : ILLMProvider
         {
             _rateLimitingService.ReleasePermit(ProviderName);
         }
+    }
+
+
+    public async Task<IAsyncEnumerable<string>> GetCompletionStreamAsync(string prompt)
+    {
+        // Generate cache key from prompt
+        var cacheKey = GenerateCacheKey(prompt);
+
+        // Try to get from cache first
+        var cachedResponse = await _cacheProvider.GetAsync<string>(cacheKey);
+        if (cachedResponse != null)
+        {
+            // For cached responses, return as a single chunk
+            return AsyncEnumerable.Singleton(cachedResponse);
+        }
+
+        // Get new completion with rate limiting
+        await _rateLimitingService.AcquirePermitAsync(ProviderName);
+        
+        // Create a StringBuilder to accumulate the full response for caching
+        var fullResponse = new StringBuilder();
+
+        // Return an async stream that accumulates the response and caches it when complete
+        async IAsyncEnumerable<string> StreamWithCaching(
+            [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                await foreach (var chunk in GetCompletionStreamInternalAsync(prompt)
+                    .WithCancellation(cancellationToken))
+                {
+                    fullResponse.Append(chunk);
+                    yield return chunk;
+                }
+
+                // Cache the complete response
+                var completeResponse = fullResponse.ToString();
+                await _cacheProvider.SetAsync(cacheKey, completeResponse);
+            }
+            finally
+            {
+                _rateLimitingService.ReleasePermit(ProviderName);
+            }
+        }
+
+        return StreamWithCaching();
     }
 
     private string GenerateCacheKey(string prompt)

--- a/src/XPlain/Services/BaseLLMProvider.cs
+++ b/src/XPlain/Services/BaseLLMProvider.cs
@@ -9,13 +9,16 @@ public abstract class BaseLLMProvider : ILLMProvider
 {
     protected readonly ICacheProvider _cacheProvider;
     protected readonly IRateLimitingService _rateLimitingService;
+    protected readonly StreamingSettings _streamingSettings;
 
     protected BaseLLMProvider(
         ICacheProvider cacheProvider,
-        IRateLimitingService rateLimitingService)
+        IRateLimitingService rateLimitingService,
+        StreamingSettings streamingSettings)
     {
         _cacheProvider = cacheProvider;
         _rateLimitingService = rateLimitingService;
+        _streamingSettings = streamingSettings;
     }
 
     public abstract string ProviderName { get; }
@@ -23,7 +26,15 @@ public abstract class BaseLLMProvider : ILLMProvider
 
     protected abstract Task<string> GetCompletionInternalAsync(string prompt);
     
-    protected abstract IAsyncEnumerable<string> GetCompletionStreamInternalAsync(string prompt);
+    /// <summary>
+    /// Gets a streaming completion from the LLM provider
+    /// </summary>
+    /// <param name="prompt">The prompt to send to the LLM</param>
+    /// <param name="cancellationToken">Token to cancel the streaming operation</param>
+    /// <returns>An async enumerable of completion chunks</returns>
+    protected abstract IAsyncEnumerable<string> GetCompletionStreamInternalAsync(
+        string prompt,
+        CancellationToken cancellationToken = default);
 
     public async Task<string> GetCompletionAsync(string prompt)
     {

--- a/src/XPlain/Services/BaseLLMProvider.cs
+++ b/src/XPlain/Services/BaseLLMProvider.cs
@@ -55,7 +55,9 @@ public abstract class BaseLLMProvider : ILLMProvider
     }
 
 
-    public async Task<IAsyncEnumerable<string>> GetCompletionStreamAsync(string prompt)
+    public async Task<IAsyncEnumerable<string>> GetCompletionStreamAsync(
+        string prompt,
+        CancellationToken cancellationToken = default)
     {
         // Generate cache key from prompt
         var cacheKey = GenerateCacheKey(prompt);
@@ -69,27 +71,37 @@ public abstract class BaseLLMProvider : ILLMProvider
         }
 
         // Get new completion with rate limiting
-        await _rateLimitingService.AcquirePermitAsync(ProviderName);
+        await _rateLimitingService.AcquirePermitAsync(ProviderName, cancellationToken: cancellationToken);
         
         // Create a StringBuilder to accumulate the full response for caching
         var fullResponse = new StringBuilder();
 
         // Return an async stream that accumulates the response and caches it when complete
         async IAsyncEnumerable<string> StreamWithCaching(
-            [EnumeratorCancellation] CancellationToken cancellationToken = default)
+            [EnumeratorCancellation] CancellationToken streamCancellationToken)
         {
             try
             {
                 await foreach (var chunk in GetCompletionStreamInternalAsync(prompt)
-                    .WithCancellation(cancellationToken))
+                    .WithCancellation(streamCancellationToken))
                 {
+                    if (streamCancellationToken.IsCancellationRequested)
+                    {
+                        // Don't cache partial responses when cancelled
+                        _rateLimitingService.ReleasePermit(ProviderName);
+                        yield break;
+                    }
+
                     fullResponse.Append(chunk);
                     yield return chunk;
                 }
 
-                // Cache the complete response
-                var completeResponse = fullResponse.ToString();
-                await _cacheProvider.SetAsync(cacheKey, completeResponse);
+                // Only cache complete responses
+                if (!streamCancellationToken.IsCancellationRequested)
+                {
+                    var completeResponse = fullResponse.ToString();
+                    await _cacheProvider.SetAsync(cacheKey, completeResponse);
+                }
             }
             finally
             {
@@ -97,7 +109,7 @@ public abstract class BaseLLMProvider : ILLMProvider
             }
         }
 
-        return StreamWithCaching();
+        return StreamWithCaching(cancellationToken);
     }
 
     private string GenerateCacheKey(string prompt)

--- a/src/XPlain/Services/ILLMProvider.cs
+++ b/src/XPlain/Services/ILLMProvider.cs
@@ -1,3 +1,4 @@
+using System.Collections.Generic;
 using System.Threading.Tasks;
 
 namespace XPlain.Services
@@ -10,6 +11,13 @@ namespace XPlain.Services
         /// <param name="prompt">The prompt to send to the LLM</param>
         /// <returns>The completion response from the LLM</returns>
         Task<string> GetCompletionAsync(string prompt);
+
+        /// <summary>
+        /// Gets a streaming completion from the LLM provider
+        /// </summary>
+        /// <param name="prompt">The prompt to send to the LLM</param>
+        /// <returns>An async enumerable of completion chunks from the LLM</returns>
+        Task<IAsyncEnumerable<string>> GetCompletionStreamAsync(string prompt);
 
         /// <summary>
         /// Gets the name of the LLM provider

--- a/src/XPlain/Services/OpenAIClient.cs
+++ b/src/XPlain/Services/OpenAIClient.cs
@@ -1,10 +1,13 @@
 using System.Net.Http.Headers;
 using System.Net.Http.Json;
+using System.Text.Json;
 using System.Text.Json.Serialization;
 using Microsoft.Extensions.Options;
 using System.Net;
 using System.Diagnostics;
 using XPlain.Configuration;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
 
 namespace XPlain.Services;
 
@@ -72,6 +75,126 @@ public class OpenAIClient : BaseLLMProvider, IOpenAIClient, IDisposable
             return response?.Choices.FirstOrDefault()?.Message.Content.Trim() ?? 
                    "No response received from the API.";
         }, 0);
+    }
+
+    protected override async IAsyncEnumerable<string> GetCompletionStreamInternalAsync(
+        string prompt)
+    {
+        var requestBody = new OpenAIRequest
+        {
+            Model = _settings.DefaultModel,
+            Messages = [
+                new OpenAIMessage
+                {
+                    Role = "user",
+                    Content = prompt
+                }
+            ],
+            MaxTokens = _settings.MaxTokenLimit,
+            Temperature = 0.7,
+            Stream = true
+        };
+
+        await foreach (var chunk in ExecuteStreamWithRetryAsync(requestBody))
+        {
+            yield return chunk;
+        }
+    }
+
+    private async IAsyncEnumerable<string> ExecuteStreamWithRetryAsync(
+        OpenAIRequest request,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        if (!_circuitBreaker.CanProcess())
+        {
+            throw new CircuitBreakerOpenException(
+                "Circuit breaker is open. Too many recent failures.");
+        }
+
+        var attempts = 0;
+        var delay = _settings.InitialRetryDelayMs;
+        Exception lastException = null;
+
+        while (attempts < _settings.MaxRetryAttempts)
+        {
+            try
+            {
+                attempts++;
+                await foreach (var chunk in ProcessStreamRequestAsync(request, cancellationToken))
+                {
+                    yield return chunk;
+                }
+                _circuitBreaker.RecordSuccess();
+                yield break;
+            }
+            catch (HttpRequestException ex) when (
+                ex.StatusCode.HasValue &&
+                RetryableStatusCodes.Contains(ex.StatusCode.Value))
+            {
+                lastException = ex;
+                _circuitBreaker.RecordFailure();
+
+                if (attempts < _settings.MaxRetryAttempts)
+                {
+                    var jitter = Random.Shared.NextDouble() * _settings.JitterFactor * delay;
+                    await Task.Delay((int)(delay + jitter), cancellationToken);
+                    delay = Math.Min(
+                        (int)(delay * _settings.BackoffMultiplier),
+                        _settings.MaxRetryDelayMs);
+                }
+            }
+            catch (Exception ex)
+            {
+                lastException = ex;
+                _circuitBreaker.RecordFailure();
+                throw;
+            }
+        }
+
+        throw new OpenAIApiException(
+            $"Stream operation failed after {attempts} attempts. Last error: {lastException?.Message}",
+            lastException);
+    }
+
+    private async IAsyncEnumerable<string> ProcessStreamRequestAsync(
+        OpenAIRequest request,
+        [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        // Acquire rate limiting permit
+        await _rateLimitingService.AcquirePermitAsync(ProviderName, 0, cancellationToken);
+        try
+        {
+            using var response = await _httpClient.PostAsJsonAsync(
+                "/v1/chat/completions",
+                request,
+                cancellationToken);
+
+            response.EnsureSuccessStatusCode();
+            
+            using var stream = await response.Content.ReadAsStreamAsync(cancellationToken);
+            using var reader = new StreamReader(stream);
+
+            while (!reader.EndOfStream)
+            {
+                var line = await reader.ReadLineAsync();
+                if (string.IsNullOrEmpty(line)) continue;
+                if (!line.StartsWith("data: ")) continue;
+                if (line == "data: [DONE]") break;
+
+                var jsonData = line.Substring(6);
+                var streamResponse = JsonSerializer.Deserialize<OpenAIStreamResponse>(jsonData);
+                
+                if (streamResponse?.Choices?.Count > 0 && 
+                    !string.IsNullOrEmpty(streamResponse.Choices[0].Delta?.Content))
+                {
+                    yield return streamResponse.Choices[0].Delta.Content;
+                }
+            }
+        }
+        finally
+        {
+            _rateLimitingService.ReleasePermit(ProviderName);
+        }
     }
 
     private async Task<OpenAIResponse> ProcessRequestAsync(
@@ -266,9 +389,30 @@ public class OpenAIChoice
     [JsonPropertyName("message")]
     public required OpenAIMessage Message { get; set; }
 
+    [JsonPropertyName("delta")]
+    public OpenAIMessage? Delta { get; set; }
+
     [JsonPropertyName("finish_reason")]
     public required string FinishReason { get; set; }
 
     [JsonPropertyName("index")]
     public required int Index { get; set; }
+}
+
+public class OpenAIStreamResponse
+{
+    [JsonPropertyName("id")]
+    public required string Id { get; set; }
+
+    [JsonPropertyName("object")]
+    public required string Object { get; set; }
+
+    [JsonPropertyName("created")]
+    public required int Created { get; set; }
+
+    [JsonPropertyName("model")]
+    public required string Model { get; set; }
+
+    [JsonPropertyName("choices")]
+    public required List<OpenAIChoice> Choices { get; set; }
 }

--- a/src/XPlain/Services/OpenAIClient.cs
+++ b/src/XPlain/Services/OpenAIClient.cs
@@ -41,7 +41,11 @@ public class OpenAIClient : BaseLLMProvider, IOpenAIClient, IDisposable
         _settings = settings.Value;
         _rateLimitingService = rateLimitingService;
 
-        _httpClient = new HttpClient
+        _httpClient = new HttpClient(
+            new StreamingHttpHandler(
+                timeout: TimeSpan.FromSeconds(30),
+                maxRetries: 3,
+                initialRetryDelay: TimeSpan.FromSeconds(1)))
         {
             BaseAddress = new Uri(_settings.ApiEndpoint)
         };
@@ -164,8 +168,9 @@ public class OpenAIClient : BaseLLMProvider, IOpenAIClient, IDisposable
         await _rateLimitingService.AcquirePermitAsync(ProviderName, 0, cancellationToken);
         try
         {
+            var streamingEndpoint = request.Stream ? "/v1/chat/completions/stream" : "/v1/chat/completions";
             using var response = await _httpClient.PostAsJsonAsync(
-                "/v1/chat/completions",
+                streamingEndpoint,
                 request,
                 cancellationToken);
 

--- a/src/XPlain/Services/OpenAIClient.cs
+++ b/src/XPlain/Services/OpenAIClient.cs
@@ -35,17 +35,14 @@ public class OpenAIClient : BaseLLMProvider, IOpenAIClient, IDisposable
     public OpenAIClient(
         IOptions<OpenAISettings> settings,
         ICacheProvider cacheProvider,
-        IRateLimitingService rateLimitingService)
-        : base(cacheProvider, rateLimitingService)
+        IRateLimitingService rateLimitingService,
+        StreamingSettings streamingSettings)
+        : base(cacheProvider, rateLimitingService, streamingSettings)
     {
         _settings = settings.Value;
         _rateLimitingService = rateLimitingService;
 
-        _httpClient = new HttpClient(
-            new StreamingHttpHandler(
-                timeout: TimeSpan.FromSeconds(30),
-                maxRetries: 3,
-                initialRetryDelay: TimeSpan.FromSeconds(1)))
+        _httpClient = new HttpClient(new StreamingHttpHandler(streamingSettings))
         {
             BaseAddress = new Uri(_settings.ApiEndpoint)
         };

--- a/src/XPlain/Services/ProgressIndicator.cs
+++ b/src/XPlain/Services/ProgressIndicator.cs
@@ -1,0 +1,82 @@
+namespace XPlain.Services;
+
+public class ProgressIndicator : IDisposable
+{
+    private readonly Timer _timer;
+    private readonly string[] _frames;
+    private int _currentFrame;
+    private bool _isFirst = true;
+    private readonly int _updateIntervalMs;
+    private bool _disposed;
+
+    public ProgressIndicator(int updateIntervalMs = 100)
+    {
+        _updateIntervalMs = updateIntervalMs;
+        _frames = new[] { "⣾", "⣽", "⣻", "⢿", "⡿", "⣟", "⣯", "⣷" };
+        _timer = new Timer(UpdateFrame, null, Timeout.Infinite, Timeout.Infinite);
+    }
+
+    public void Start()
+    {
+        if (_disposed) return;
+        _timer.Change(0, _updateIntervalMs);
+    }
+
+    public void Stop()
+    {
+        if (_disposed) return;
+        _timer.Change(Timeout.Infinite, Timeout.Infinite);
+        ClearLastFrame();
+    }
+
+    private void UpdateFrame(object? state)
+    {
+        if (_disposed || Console.IsOutputRedirected) return;
+
+        try
+        {
+            if (_isFirst)
+            {
+                _isFirst = false;
+                Console.Write(_frames[0]);
+            }
+            else
+            {
+                Console.Write("\b");
+                Console.Write(_frames[_currentFrame]);
+            }
+
+            _currentFrame = (_currentFrame + 1) % _frames.Length;
+        }
+        catch
+        {
+            // Ignore console errors
+        }
+    }
+
+    private void ClearLastFrame()
+    {
+        if (Console.IsOutputRedirected) return;
+
+        try
+        {
+            if (!_isFirst)
+            {
+                Console.Write("\b \b");
+            }
+        }
+        catch
+        {
+            // Ignore console errors
+        }
+    }
+
+    public void Dispose()
+    {
+        if (_disposed) return;
+        
+        _disposed = true;
+        Stop();
+        _timer.Dispose();
+    }
+}

--- a/src/XPlain/Services/StreamingHttpHandler.cs
+++ b/src/XPlain/Services/StreamingHttpHandler.cs
@@ -13,9 +13,9 @@ public class StreamingHttpHandler : DelegatingHandler
 
     public StreamingHttpHandler(StreamingSettings settings)
     {
-        _timeout = timeout;
-        _maxRetries = maxRetries;
-        _initialRetryDelay = initialRetryDelay ?? TimeSpan.FromSeconds(1);
+        _timeout = TimeSpan.FromSeconds(settings.StreamingTimeoutSeconds);
+        _maxRetries = settings.MaxStreamingRetries;
+        _initialRetryDelay = TimeSpan.FromMilliseconds(settings.InitialRetryDelayMs);
         InnerHandler = new HttpClientHandler
         {
             AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate

--- a/src/XPlain/Services/StreamingHttpHandler.cs
+++ b/src/XPlain/Services/StreamingHttpHandler.cs
@@ -11,10 +11,7 @@ public class StreamingHttpHandler : DelegatingHandler
     private readonly int _maxRetries;
     private readonly TimeSpan _initialRetryDelay;
 
-    public StreamingHttpHandler(
-        TimeSpan timeout,
-        int maxRetries = 3,
-        TimeSpan? initialRetryDelay = null)
+    public StreamingHttpHandler(StreamingSettings settings)
     {
         _timeout = timeout;
         _maxRetries = maxRetries;

--- a/src/XPlain/Services/StreamingHttpHandler.cs
+++ b/src/XPlain/Services/StreamingHttpHandler.cs
@@ -1,0 +1,72 @@
+using System.Net;
+
+namespace XPlain.Services;
+
+/// <summary>
+/// HTTP message handler that provides better stability for streaming responses
+/// </summary>
+public class StreamingHttpHandler : DelegatingHandler
+{
+    private readonly TimeSpan _timeout;
+    private readonly int _maxRetries;
+    private readonly TimeSpan _initialRetryDelay;
+
+    public StreamingHttpHandler(
+        TimeSpan timeout,
+        int maxRetries = 3,
+        TimeSpan? initialRetryDelay = null)
+    {
+        _timeout = timeout;
+        _maxRetries = maxRetries;
+        _initialRetryDelay = initialRetryDelay ?? TimeSpan.FromSeconds(1);
+        InnerHandler = new HttpClientHandler
+        {
+            AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate
+        };
+    }
+
+    protected override async Task<HttpResponseMessage> SendAsync(
+        HttpRequestMessage request,
+        CancellationToken cancellationToken)
+    {
+        var retryCount = 0;
+        var delay = _initialRetryDelay;
+
+        while (true)
+        {
+            try
+            {
+                using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                cts.CancelAfter(_timeout);
+
+                var response = await base.SendAsync(request, cts.Token);
+                
+                if ((int)response.StatusCode >= 500 && retryCount < _maxRetries)
+                {
+                    await Task.Delay(delay, cancellationToken);
+                    delay = TimeSpan.FromMilliseconds(delay.TotalMilliseconds * 2);
+                    retryCount++;
+                    continue;
+                }
+
+                response.EnsureSuccessStatusCode();
+                return response;
+            }
+            catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
+            {
+                if (retryCount >= _maxRetries)
+                    throw new TimeoutException($"Request timed out after {_timeout.TotalSeconds} seconds");
+
+                await Task.Delay(delay, cancellationToken);
+                delay = TimeSpan.FromMilliseconds(delay.TotalMilliseconds * 2);
+                retryCount++;
+            }
+            catch (HttpRequestException ex) when (retryCount < _maxRetries)
+            {
+                await Task.Delay(delay, cancellationToken);
+                delay = TimeSpan.FromMilliseconds(delay.TotalMilliseconds * 2);
+                retryCount++;
+            }
+        }
+    }
+}

--- a/tests/XPlain.Tests/XPlain.Tests.csproj
+++ b/tests/XPlain.Tests/XPlain.Tests.csproj
@@ -1,29 +1,29 @@
-&lt;Project Sdk="Microsoft.NET.Sdk"&gt;
+<Project Sdk="Microsoft.NET.Sdk">
 
-  &lt;PropertyGroup&gt;
-    &lt;TargetFramework&gt;net7.0&lt;/TargetFramework&gt;
-    &lt;ImplicitUsings&gt;enable&lt;/ImplicitUsings&gt;
-    &lt;Nullable&gt;enable&lt;/Nullable&gt;
-    &lt;IsPackable&gt;false&lt;/IsPackable&gt;
-    &lt;IsTestProject&gt;true&lt;/IsTestProject&gt;
-  &lt;/PropertyGroup&gt;
+  <PropertyGroup>
+    <TargetFramework>net7.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <IsPackable>false</IsPackable>
+    <IsTestProject>true</IsTestProject>
+  </PropertyGroup>
 
-  &lt;ItemGroup&gt;
-    &lt;PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.6.0" /&gt;
-    &lt;PackageReference Include="xunit" Version="2.4.2" /&gt;
-    &lt;PackageReference Include="xunit.runner.visualstudio" Version="2.4.5"&gt;
-      &lt;IncludeAssets&gt;runtime; build; native; contentfiles; analyzers; buildtransitive&lt;/IncludeAssets&gt;
-      &lt;PrivateAssets&gt;all&lt;/PrivateAssets&gt;
-    &lt;/PackageReference&gt;
-    &lt;PackageReference Include="coverlet.collector" Version="6.0.0"&gt;
-      &lt;IncludeAssets&gt;runtime; build; native; contentfiles; analyzers; buildtransitive&lt;/IncludeAssets&gt;
-      &lt;PrivateAssets&gt;all&lt;/PrivateAssets&gt;
-    &lt;/PackageReference&gt;
-    &lt;PackageReference Include="Moq" Version="4.18.4" /&gt;
-  &lt;/ItemGroup&gt;
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.6.0" />
+    <PackageReference Include="xunit" Version="2.4.2" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="coverlet.collector" Version="6.0.0">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="Moq" Version="4.18.4" />
+  </ItemGroup>
 
-  &lt;ItemGroup&gt;
-    &lt;ProjectReference Include="..\..\src\XPlain\XPlain.csproj" /&gt;
-  &lt;/ItemGroup&gt;
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\XPlain\XPlain.csproj" />
+  </ItemGroup>
 
-&lt;/Project&gt;
+</Project>

